src/Microsoft.AspNetCore.Analyzer.Testing/DiagnosticVerifier.cs

```diff
@@ -0,0 +1,189 @@
+        {
+            var projects = new HashSet<Project>();
+            foreach (var document in documents)
+            {
+                projects.Add(document.Project);
+            }
+
+            var diagnostics = new List<Diagnostic>();
+            foreach (var project in projects)
+            {
+                var compilation = await project.GetCompilationAsync();
+
+                // Enable any additional diagnostics
+                var options = compilation.Options;
+                if (additionalEnabledDiagnostics.Length > 0)
+                {
+                    options = compilation.Options
+                        .WithSpecificDiagnosticOptions(
+                            additionalEnabledDiagnostics.ToDictionary(s => s, s => ReportDiagnostic.Info));
+                }
+
+                var compilationWithAnalyzers = compilation
+                    .WithOptions(options)
+                    .WithAnalyzers(ImmutableArray.Create(analyzer));
+
+                var diags = await compilationWithAnalyzers.GetAllDiagnosticsAsync();
+
+                Assert.DoesNotContain(diags, d => d.Id == "AD0001");
+
+                // Filter out non-error diagnostics not produced by our analyzer
+                // We want to KEEP errors because we might have written bad code. But sometimes we leave warnings in to make the
+                // test code more convenient
+                diags = diags.Where(d => d.Severity == DiagnosticSeverity.Error || analyzer.SupportedDiagnostics.Any(s => s.Id.Equals(d.Id))).ToImmutableArray();
+
+                foreach (var diag in diags)
+                {
+                    if (diag.Location == Location.None || diag.Location.IsInMetadata)
+                    {
+                        diagnostics.Add(diag);
+                    }
+                    else
+                    {
+                        foreach (var document in documents)
+                        {
+                            var tree = await document.GetSyntaxTreeAsync();
+                            if (tree == diag.Location.SourceTree)
+                            {
+                                diagnostics.Add(diag);
+                            }
+                        }
+                    }
+                }
+            }
+
+            return diagnostics.OrderBy(d => d.Location.SourceSpan.Start).ToArray();
+        }
+
+        /// <summary>
+        /// Given an array of strings as sources and a language, turn them into a project and return the documents and spans of it.
+        /// </summary>
+        /// <param name="sources">Classes in the form of strings</param>
+        /// <returns>An array of Documents produced from the sources.</returns>
+        private Document[] GetDocuments(string[] sources)
+        {
+            var project = CreateProject(sources);
+            var documents = project.Documents.ToArray();
+
+            if (sources.Length != documents.Length)
```

Comment:
Do you know when this is true? If so, could you include what the test author would need to do to remedy it?